packages/go_router/lib/src/delegate.dart

```diff
@@ -146,15 +160,40 @@ class GoRouterDelegate extends RouterDelegate<RouteMatchList>
     return true;
   }
 
-  /// Replaces the top-most page of the page stack with the given one.
+  /// Replaces the top-most page of the page stack with the given one. The page
+  /// key of the new page will always be different from the old one.
   ///
   /// See also:
   /// * [push] which pushes the given location onto the page stack.
+  /// * [replace] which replaces the top-most page of the page stack but keeps
+  ///   the page key when possible.
   void pushReplacement(RouteMatchList matches) {
+    assert(matches.last.route is! ShellRoute);
     _matchList.remove(_matchList.last);
     push(matches); // [push] will notify the listeners.
   }
 
+  /// Replaces the top-most page of the page stack with the given one and reuse
```

Comment:
This doc should not talk about page key since that is implementation detail. Instead it should talk about the visual effect of calling this method. Same thing for pushReplacement

Reply:
This seems not addressed yet?

Reply:
@chunhtai I could write in the doc that `pushReplacement` animates to the new page while `replace` does not.
But it is not only that, `pushReplacement` doesn't preserve the state of the widget tree while `replace` does. I think this is an important aspect that needs to be said and that's why I'm mentioning the page key.

Having said this, what do you think I should write?

Reply:
You should definitely mentioned animation/page transition and that the it will be treated as same page and state will be preserved. I am ok if either you keep the page key info here or not.

Reply:
This doesn't seem to be addressed yet